File: EstruturasLineares/ArrayListJava/ArrayList.java
package EstruturasLineares.ArrayListJava;

public class ArrayList<T> {
    // =====PRIVATE=====//
    private static final int DOWNGRADE = 4;

    private T[] list;
    private int length = 0;
    private Boolean dynamic = true;

    // =====CONSTRUCTORS=====//
    public ArrayList() {
        this(10);
    }

    @SuppressWarnings("unchecked")
    public ArrayList(int cap) {
        this.setList((T[]) new Object[cap]);
    }

    @SuppressWarnings("unchecked")
    public ArrayList(T... arr) {
        this(true, arr);
    }

    @SuppressWarnings("unchecked")
    public ArrayList(Boolean dynamic, T... arr) {
        this.setDynamic(dynamic);
        this.setList((T[]) new Object[arr.length]);
        this.addAll(arr);
    }

    // =====GET=====//
    public T[] getList() {
        return this.list;
    }

    public int size() {
        return this.length;
    }

    public int capacity() {
        return this.list.length;
    }

    public T get(int idx) {
        if (idx < 0 || idx > this.size())
            throw new IndexOutOfBoundsException();
        return this.list[idx];
    }

    public Boolean isDynamic() {
        return this.dynamic;
    }

    // =====SET=====//
    // -----PRIVATE-----//
    private void setList(T[] l) {
        this.list = l;
    }

    private void setLength(int length) {
        this.length = length;
    }

    private void inc() {
        this.length++;
    }

    private void dec() {
        this.length--;
    }

    private void set(T obj, int p) {
        this.list[p] = obj;
    }

    // -----PUBLIC-----//
    public void setDynamic(Boolean value) {
        this.dynamic = value;
    }

    // =====METODOS=====//
    public Boolean isEmpty() {
        return this.size() == 0;
    }

    public Boolean isFull() {
        return this.size() == this.capacity();
    }

    // =====GROW=====//
    @SuppressWarnings("unchecked")
    private void grow() {
        T[] arr = (T[]) new Object[this.capacity() * 2];
        for (int i = 0; i < this.size(); i++)
            arr[i] = this.get(i);

        this.setList(arr);
    }

    // =====SHRINK=====//
    @SuppressWarnings("unchecked")
    private void shrink() {
        T[] arr = (T[]) new Object[this.capacity() / 2];
        for (int i = 0; i < this.size(); i++)
            arr[i] = this.get(i);

        this.setList(arr);
    }

    // =====ADD=====//
    public void add(T obj) {
        if (this.isFull()) {
            if (this.isDynamic())
                this.grow();
            else
                throw new FullArrayListException();
        }

        set(obj, this.size());
        inc();
    }

    public void add(T obj, int p) {
        if (this.isFull()) {
            if (this.isDynamic())
                this.grow();
            else
                throw new FullArrayListException();
        }
        if (p < 0 || p >= this.size())
            throw new IndexOutOfBoundsException();

        for (int i = this.size(); i > p; i--)
            set(this.get(i - 1), i);
        set(obj, p);
        inc();
    }

    public void addAll(T[] arr) {
        if ((this.capacity() - this.size()) < arr.length) {
            if (this.isDynamic())
                this.grow();
            else
                throw new FullArrayListException();
        }

        for (int i = 0; i < arr.length; i++)
            this.add(arr[i]);
    }

    // =====INSERT=====//
    public void insert(T obj) {
        if (this.isFull()) {
            if (this.isDynamic())
                this.grow();
            else
                throw new FullArrayListException();
        }

        for (int i = this.size(); i > 0; i--)
            set(this.get(i - 1), i);
        set(obj, 0);
        inc();
    }

    // =====POP=====//
    public T pop() {
        if (this.isEmpty())
            throw new EmptyArrayListException();

        dec();
        if (this.isDynamic() && this.size() <= this.capacity() / DOWNGRADE)
            this.shrink();
        return this.get(this.size());
    }

    public T pop(int idx) {
        if (this.isEmpty())
            throw new EmptyArrayListException();
        if (idx < 0 || idx >= this.size())
            throw new IndexOutOfBoundsException();

        T aux = this.get(idx);
        for (int i = idx; i < this.size() - 1; i++)
            this.set(this.get(i + 1), i);
        dec();
        if (this.isDynamic() && this.size() <= this.capacity() / DOWNGRADE)
            this.shrink();
        return aux;
    }

    public T pop(T obj) {
        if (this.isEmpty())
            throw new EmptyArrayListException();

        int idx = search(obj);
        T aux = null;
        if (idx >= 0) {
            aux = this.get(idx);
            for (int i = idx; i < this.size() - 1; i++)
                this.set(this.get(i + 1), i);
            dec();
            if (this.isDynamic() && this.size() <= this.capacity() / DOWNGRADE)
                this.shrink();
        }
        return aux;
    }

    // =====REMOVE=====//
    public T remove() {
        if (this.isEmpty())
            throw new EmptyArrayListException();

        T aux = this.get(0);
        for (int i = 0; i < this.size() - 1; i++)
            this.set(this.get(i + 1), i);
        dec();
        if (this.isDynamic() && this.size() <= this.capacity() / DOWNGRADE)
            this.shrink();
        return aux;
    }

    // =====SEARCH=====//
    public int search(T obj) {
        if (obj == null) {
            for (int i = 0; i < this.size(); i++)
                if (this.get(i) == null)
                    return i;
        } else {
            for (int i = 0; i < this.size(); i++)
                if (obj.equals(this.get(i)))
                    return i;
        }

        return -1;
    }

    // =====CONTAINS=====//
    public Boolean contains(T obj) {
        Boolean value = false;
        int i = 0;
        if (obj == null) {
            while (i < this.size() && !value) {
                value = this.get(i) == null;
                i++;
            }
        } else {
            while (i < this.size() && !value) {
                value = obj.equals(this.get(i));
                i++;
            }
        }

        return value;
    }

    // =====CLEAR=====//
    public void clear() {
        this.setLength(0);
        for(int i = 0; i < this.capacity(); i++)
            this.set(null, i);
    }

    // =====OVERRIDE=====//
    @Override
    public String toString() {
        String s = "[";
        for (int i = 0; i < this.size(); i++) {
            s += this.get(i);
            if (i != this.size() - 1)
                s += ", ";
        }
        s += "]";
        return s;
    }

}

File: EstruturasLineares/ArrayListC/comp/ArrayList.c
//=====INCLUDE=====//
#include "../header/ArrayList.h" 

//=====METODOS=====//
//-----CONSTRUCTORS-----//
void alocateStrs(Strings* const arr, int i) {
    arr->strs = (char**)calloc(i, sizeof(char*));
    arr->cap = i;
    arr->len = 0;
}

void freeStrs(Strings* arr) {
    free(arr->strs);
}

//-----ISEMPTY-----//
bool isEmpty(Strings* const arr) {
    return arr->len == 0;
}

//-----ISFULL-----//
bool isFull(Strings* const arr) {
    return arr->len == arr->cap;
}

//-----GROW-----//
void grow(Strings* arr) {
    Strings newArr;
    alocateStrs(&newArr, arr->cap * 2);
    
    for(int i = 0; i < arr->len; i++)
        newArr.strs[i] = arr->strs[i];

    free(arr->strs);
    arr->cap = newArr.cap;
    arr->strs = newArr.strs;
}

//-----SHRINK-----//
void shrink(Strings* arr) {
    Strings newArr;
    alocateStrs(&newArr, arr->cap / 2);
    
    for(int i = 0; i < arr->len; i++)
        newArr.strs[i] = arr->strs[i];
    
    free(arr->strs);
    arr->cap = newArr.cap;
    arr->strs = newArr.strs;
}

//-----ADD-----//
void add(Strings* const arr, char* const s) {
    if(isFull(arr)) grow(arr);
    arr->strs[arr->len] = s;
    arr->len++;
}

void addp(Strings* const arr, int p, char* const s) {
    if(isFull(arr)) grow(arr);
    for(int i = arr->len; i > p; i--) 
        arr->strs[i] = arr->strs[i - 1];
    
    arr->strs[p] = s;
    arr->len++;
}

//-----INSERT-----//
void insert(Strings* const arr, char* const s) {
    if(isFull(arr)) grow(arr);
    for(int i = arr->len; i > 0; i--) 
        arr->strs[i] = arr->strs[i - 1];

    arr->strs[0] = s;
    arr->len++;
}

//-----POP-----//
char* pop(Strings* const arr) {
    char* s = NULL;
    if(!isEmpty(arr)) {
        arr->len--;
        s = arr->strs[arr->len];

        if(arr->len <= (arr->cap / DOWNGRADE)) shrink(arr);
    }
    return s;
}

char* popp(Strings* const arr, int p) {
    char* s = NULL;
    if(!isEmpty(arr)) {
        s = arr->strs[p];
        for(int i = p; i < arr->len - 1; i++) 
            arr->strs[i] = arr->strs[i + 1];

        arr->len--;
        if(arr->len <= (arr->cap / DOWNGRADE)) shrink(arr);
    }
    return s;
}

char* popObj(Strings* const arr, char* s) {
    if(!isEmpty(arr)) 
        return popp(arr, search(arr, s));
    return NULL;
}

//-----REMOVE-----//
char* removee(Strings* const arr) {
    char* s = NULL;
    if(!isEmpty(arr)) {
        s = arr->strs[0];
        for(int i = 0; i < arr->len - 1; i++) 
        	arr->strs[i] = arr->strs[i + 1];
        
        arr->len--;
        if(arr->len <= (arr->cap / DOWNGRADE)) shrink(arr);
    }
    return s;
}

//-----SEARCH-----//
int search(Strings* const arr, char* const s) {
    if(s == NULL) {
        for(int i = 0; i < arr->len; i++) 
            if(arr->strs[i] == NULL) return i;
    } else {
        for(int i = 0; i < arr->len; i++)
            if(equals(arr->strs[i], s)) return i;
    }
    return -1;
}

//-----CONTAINS-----//
bool contains(Strings* const arr, char* const s) {
    bool value = false;
    if(s == NULL) {
        for(int i = 0; i < arr->len && !value; i++) 
            value = arr->strs[i] == NULL; 
    } else {
        for(int i = 0; i < arr->len && !value; i++) 
            value = equals(arr->strs[i], s);
    }
    return value;
}

//=====QUICKSORT=====//
void quickSort(Strings* const arr, int first, int last) {
    if(first < last) {
        int p = partition(arr, first, last);

        quickSort(arr, first, p - 1);
        quickSort(arr, p, last);
    }
}

int partition(Strings* const arr, int first, int last) {
    int i = first,
        j = last;

    char* p = arr->strs[(first + last) / 2];

    while(i <= j) {
        while(compareTo(arr->strs[i], p) < 0) i++;
        while(compareTo(arr->strs[j], p) > 0) j--;
        if(i <= j) {
            swap(arr, i, j);
            i++;
            j--;
        }
    }
    
    return i;
}

void swap(Strings* const arr, int a, int b) {
    char* s = arr->strs[a];
    arr->strs[a] = arr->strs[b];
    arr->strs[b] = s;
}

//=====PRINT=====//
void printStrs(Strings* arr) {
    printf("[");
    for(int i = 0; i < arr->len; i++) {
        printf("%s", arr->strs[i]);
        if(i != arr->len - 1) printf(", ");
    } 
    puts("]");
}

//=====ISBOOL=====//
bool isNumber(char c) {
    return (int)c >= 48 && (int)c <= 57;
}

bool isFloat(char* const s) {
    bool value = true;
    for(int i = 0, j = 0; i < length(s) && value; i++) {
        if(s[i] == ',' || s[i] == '.') j++; 
        else value = isNumber(s[i]);
        if(value) value = j <= 1; 
    } 
    return value;
}

bool equals(const char* const s1, const char* const s2) {
    int len1 = length(s1),
        len2 = length(s2);
    bool value = len1 == len2;

    for(int i = 0; i < len1 && value; i++) 
        value = s1[i] == s2[i];
    return value;
}

//=====PARSES=====//
int parseInt(char* const s) {
    int num,
        len = length(s);

    for(int i = len - 1, j = 1; i >= 0; i--, j *= 10)  
        if(isNumber(s[i]))
            num += ((int)s[i] - 48) * j;
    return num;
}

//=====COMPARE=====//
int compareTo(char* const s1, char* const s2) {
    int len1 = length(s1),
        len2 = length(s2),
        min = (len1 < len2) ? len1 : len2;
    
    for(int i = 0; i < min; i++) 
        if(s1[i] != s2[i]) 
            return s1[i] - s2[i];

    return len1 - len2;
}

//=====COUNTERS=====//
size_t length(const char* const s) {
    size_t i = 0;
    if(s != NULL)
        while(s[i] != '\0') i++;
    return i;
}

//=====MATH=====//
double expfloat(double b, int e) {
    double res = 1;
    for(int i = 0; i < e; i++) res *= b;
    return res;
}

int expinteger(int b, int e) {
    int res = 1;
    for(int i = 0; i < e; i++) res *= b;
    return res;
}
